binary_search_tree/tree.py

```diff
@@ -14,44 +14,162 @@ class Tree:
+        else:
+            current_node.right = self.add_helper(current_node.right, key, value)
+        return current_node
+
+    # Time Complexity: O(n), O(h)
+    # Space Complexity: O(1)
     def add(self, key, value = None):
-        pass
+        if self.root == None:
+            self.root = TreeNode(key, value)
+        else:
+            self.add_helper(self.root, key, value)
 
-    # Time Complexity: 
-    # Space Complexity: 
+    # Time Complexity: O(n), O(h)
+    # Space Complexity: O(1)
     def find(self, key):
-        pass
+        if self.root == None:
+            return None
+
+        current = self.root
+        while current != None:
+            if current.key == key:
+                return current.value
+            elif key < current.key:
+                current = current.left
+            else:
+                current = current.right
+
+        return None
 
-    # Time Complexity: 
-    # Space Complexity: 
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)
     def inorder(self):
-        pass
+        tree_array=[]
+        
+        if self.root == None:
+            return tree_array
+
+        return self.helper_inorder(self.root, tree_array)
+
+    def helper_inorder(self, node, tree_array):
+        if node == None:
+            return 
+
+        self.helper_inorder(node.left, tree_array)
+        tree_array.append({
+            'key': node.key,
+            'value': node.value
+        })
+        self.helper_inorder(node.right, tree_array)
 
-    # Time Complexity: 
-    # Space Complexity:     
+        return tree_array
+
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)
     def preorder(self):
-        pass
+        tree_array = []
+
+        if self.root == None:
+            return tree_array
+
+        return self.helper_preorder(self.root, tree_array)
+
+    def helper_preorder(self, node, tree_array):
+        if node == None:
+            return
+
+        tree_array.append({
+            'key': node.key,
+            'value': node.value
+        })
+        self.helper_preorder(node.left, tree_array)
+        self.helper_preorder(node.right, tree_array)
+        
+        return tree_array
+
 
-    # Time Complexity: 
-    # Space Complexity:     
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)
     def postorder(self):
-        pass
+        tree_array = []
 
-    # Time Complexity: 
-    # Space Complexity:     
+        if self.root == None:
+            return tree_array
+
+        return self.helper_postorder(self.root, tree_array)
+
+    def helper_postorder(self, node, tree_array):
+        if node == None:
+            return 
+
+        self.helper_postorder(node.left, tree_array)
+        self.helper_postorder(node.right, tree_array)
+        tree_array.append({
+            'key': node.key,
+            'value': node.value
+        })
+
+        return tree_array
+
+
+    # Time Complexity: O(n)
+    # Space Complexity: O(1) 
     def height(self):
```

Comment:
👍 However since the helper function is recursive the space complexity is O(log n) or O(h)